R/assign.R

```diff
@@ -2,18 +2,51 @@
 sdtm_assign <- function(raw_dat,
                         raw_var,
                         tgt_var,
-                        tgt_dat = raw_dat,
-                        by = NULL,
+                        tgt_dat,
+                        by_var = NULL,
                         ct = NULL,
                         cl = NULL) {
 
   # TODO: Assertions.
 
-  raw_dat |>
-    dplyr::right_join(y = tgt_dat, by = by) |>
-    dplyr::mutate("{tgt_var}" := ct_map(!!rlang::sym(raw_var), ct = ct, cl = cl)) |>
-    dplyr::select(-rlang::sym(raw_var)) |>
-    dplyr::relocate(tgt_var, .after = dplyr::last_col())
+  # When target dataset and  by_var variables are provided,
+  # the tar_var is added to the input target dataset.
+  # This if block deals when ct is not provided.
+  if((!is.null(by_var) & is.null(ct))){
+    tgt_dat_out <- raw_dat |>
+      dplyr::right_join(y = tgt_dat, by = by_var) |>
+      dplyr::mutate("{tgt_var}" := !!rlang::sym(raw_var)) |>
+      dplyr::select(-rlang::sym(raw_var)) |>
+      dplyr::relocate(tgt_var, .after = dplyr::last_col())
+  } else if (is.null(by_var) & is.null(ct)) {
+    # When target dataset and  by_var variables are NOT provided,
+    # the tgt_dat_out is created with tar_var & oak_id_vars.
+    # This if block deals when ct is not provided.
+    tgt_dat_out <- raw_dat |>
+      dplyr::select(oak_id, raw_source, patient_number,rlang::sym(raw_var)) |>
+      dplyr::mutate("{tgt_var}" := !!rlang::sym(raw_var)) |>
+      dplyr::select(-rlang::sym(raw_var))
+  } else if (is.null(by_var) & !is.null(ct)) {
+    # When target dataset and  by_var variables are NOT provided,
+    # the tgt_dat_out is created with tar_var & oak_id_vars.
+    # This if block deals when ct is provided.
+    tgt_dat_out <- raw_dat |>
+      dplyr::select(oak_id, raw_source, patient_number,rlang::sym(raw_var)) |>
+      dplyr::mutate("{tgt_var}" := ct_map(!!rlang::sym(raw_var), ct = ct, cl = cl)) |>
+      dplyr::select(-rlang::sym(raw_var)) |>
+      dplyr::relocate(tgt_var, .after = dplyr::last_col())
+  } else if (!is.null(by_var) & !is.null(ct)) {
+    # When target dataset and  by_var variables are provided,
+    # the tar_var is added to the input target dataset.
+    # This if block deals when ct is provided.
+    tgt_dat_out <- raw_dat |>
+      dplyr::right_join(y = tgt_dat, by = by_var) |>
+      dplyr::mutate("{tgt_var}" := ct_map(!!rlang::sym(raw_var), ct = ct, cl = cl)) |>
+      dplyr::select(-rlang::sym(raw_var)) |>
+      dplyr::relocate(tgt_var, .after = dplyr::last_col())
+  }
```

Comment:
Are you sure you want to have the four if-clauses for the four combinations? The cases with and without terminology were already handled by `ct_map()` seamlessly. And couldn't you make `c("oak_id", "raw_source", "patient_number")` the default value of `by_var`, and get the code again with four lines of code? :)

Reply:
ct_map() perfectly handles both cases. From an SDTM perspective, when the algorithm is assign_no_ct, it is unnecessary to pass it on to the ct_map() function. Although you have handled it beautifully, the user will be puzzled as to why we are doing it. So I thought, it will be good to handle these four cases separately as these are separate algorithms.

We can make c("oak_id", "raw_source", "patient_number") the default value of by_var. Also, the user could add additional parameters. We need to discuss how to handle it.